enhancements/authentication/custom-scc-preemption-prevention.md

```diff
@@ -0,0 +1,260 @@
+for update events.
+
+If the requested SCC does not exist in the system, admission fails with an error.
+
+If the requested SCC does exist, any other SCC available in the system is ignored
+and the requested SCC is the only one being considered for admission.
+
+If the SCC is applicable to the workload, it gets/is used,
+else admission fails with an error.
+
+Pros:
+- The required and pinned SCC is deterministic.
+- No other SCC in the system is taken into account.
+
+Cons:
+- If the required SCC is not applicable, the workload fails to be created/admitted.
+This can be detected in the existing e2e suite.
+- All core OpenShift workloads need to set that annotation in their deployment artifacts.
+This implies that, after the change merges, many repositories have to be touched.
+
+An existing implementation is available at https://github.com/openshift/apiserver-library-go/pull/90.
+
+### Workflow Description
+
+Opting into this feature for core OpenShift workloads is straight-forward,
+with just an annotation needed to be set in the deployment manifest.
+
+### Variation (optional)
+
+Another variation for both proposed suggestions is to accept a comma-separated list of SCCs.
+This could be helpful in case additional SCCs are going to be introduced and the deployment manifest
+needs to be compatible with both old and new applicable SCCs.
+
+However, as adding a list further adds complexity, this suggestion is being omitted as an variation only
+with the potential of being introduced as a backwards-compatible change in the future.
+
+### API Extensions
+
+A new annotation `openshift.io/required-scc` is introduced.
+
+### Implementation Details/Notes/Constraints [optional]
+
+N/A
+
+### Risks and Mitigations
+
+1. OpenShift workloads that are not applicable to the requested SCC might fail to be admitted.
+Mitigation: The existing e2e test suite in OpenShift will give very early signals of failed admission attempts
+as the workload will be unschedulable.
+
+### Drawbacks
+
+This proposal adds yet another additional complexity on top of the existing SCC logic
+which also already interleaves with pod security admission.
+The mitigation here is to reduce complexity by favoring the "hard protection" variant.
+
+## Design Details
+
+### Open Questions [optional]
+
+1. There might be benefit to limit the admission plugin to reconcile the new annotation in OpenShift namespaces only.
+Drawback: OpenShift workloads might be missed who wants to opt into this functionality.
+History shows that identifying "core OpenShift namespaces" is tricky.
+
+### Test Plan
+
+The existing OpenShift e2e test suite is assumed to be sufficient to verify correctness.
+Further, an additional e2e test will be implemented that verifies that a more restrictive and highest priority
+custom SCC will not preempt a required SCC for selected workloads.
+If workloads opt into the annotation and force an SCC that is not applicable, admission will fail immediately.
```

Comment:
presumably this should be encoded as an e2e test

Reply:
good idea, I'll make sure to add this to the existing e2e test PR https://github.com/openshift/origin/pull/28048 👍